SYCL/ESIMD/api/functional/mutator.hpp

```diff
@@ -0,0 +1,94 @@
+//===-- mutator.hpp - This file provides common function and classes to mutate
+//      reference data. ---------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides common function and classes to mutate reference data.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "value.hpp"
+#include <sycl/sycl.hpp>
+
+#include <algorithm>
+#include <vector>
+
+namespace esimd_test::api::functional {
+
+// In some test cases it’s possible to pass a reference data as the input data
+// directly without any modification. For example, if we check some copy
+// constructor or memory move operation it’s OK to have any data values as the
+// input ones. But in most cases we need to consider the possibility of UB for
+// C++ operations due to modification of input values.
+//
+// Mutation mechanism covers such requirement.
+// The mutator namespace is intended to store such mutators alongside with
+// the generic ones provided below.
+namespace mutator {
+
+// Replace specific reference values to the bigger ones, so we can safely
+// substract `val` later.
+template <typename T> class For_subtraction {
+  T m_value;
+
+public:
+  For_subtraction(T val)
+      : m_value((assert(val > 0 && "Invalid value."), val)) {}
+
+  void operator()(T &val) {
+    if constexpr (std::is_signed_v<T>) {
+      if (val < value<T>::lowest() + m_value) {
+        val = value<T>::lowest() + m_value;
+      }
+    }
+  }
+};
+
+// Replace specific reference values to the smaller ones, so we can safely add
+// `val` later.
+template <typename T> class For_addition {
+  T m_value;
+
+public:
+  For_addition(T val) : m_value((assert(val > 0 && "Invalid value."), val)) {}
+
+  void operator()(T &val) {
+    if constexpr (std::is_signed_v<T> || std::is_same_v<T, sycl::half>) {
+      if (val > value<T>::max() - m_value) {
+        val = value<T>::max() - m_value;
+      }
+    }
+  }
+};
+
+// Replace specific reference values to the divided ones, so we can safely
+// multiply to `val` later.
+template <typename T> class For_division {
```

Comment:
```suggestion
// Replace specific reference values to the divided ones, so we can safely
// multiply to `val` later.
template <typename T> class For_multiplication {
```
given the description above. Looks like the logic should be fixed therefore

Reply:
This comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/789/commits/4d4d19242eb873b792da11f9a118dd5ef6e9c6cd#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R93)

Reply:
Please, update the logic. Let's for example have `using T = int`, `m_value == 2` and `val == value<int>::max()` in `operator()`.  
Looks like we don't need to touch `denorm_min` at all.

Reply:
The logic has been updated [here](https://github.com/intel/llvm-test-suite/pull/789/commits/41df5e531f17caed6253334af281f97f021058d3#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74L101-R106)